emacs/config.org
#+title: Emacs do Zero (EDZ)
#+author: Zé Neto

* Preparação
   - Instalar o Emacs
   - Criar arquivo =~/.config/emacs/init.el=
   - Criar o arquivo =~/.config/emacs/config.org=

** Conteúdo de =init.el=
#+begin_example
(org-babel-load-file
  (expand-file-name "config.org" user-emacs-directory))
#+end_example

** Configurações gerais
#+begin_src emacs-lisp
(org-indent-mode t)

;;(setq org-edit-src-content-indentation 0) ; Zera a margem dos blocos
;;(electric-indent-mode -1)                 ; Desliga a indentação automática
#+end_src

** Desabilitar tela de início (scratch)
#+begin_src emacs-lisp
(setq inhibit-startup-message t)
#+end_src

** Configuração da interface (UI)
#+begin_src emacs-lisp
(menu-bar-mode -1)            ; no menu bar
(tool-bar-mode -1)            ; no tools bar
(scroll-bar-mode -1)          ; no scroll bars
(tooltip-mode -1)             ; no tooltips
(set-fringe-mode 10)          ; frame edges set to 10px
(column-number-mode 1)        ; modeline shows column number
(save-place-mode 1)           ; remember cursor position
(recentf-mode 1)              ; remember recent files
(savehist-mode 1)             ; enable history saving
#+end_src

** Numeração de linhas
#+begin_src emacs-lisp
(global-display-line-numbers-mode 1) 
(setq display-line-numbers-type 'relative) 
(dolist (mode '(org-mode-hook
                vterm-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))
#+end_src

** Usar espaços em vez de tabulações
#+begin_src emacs-lisp
(setq-default indent-tabs-mode nil)
#+end_src

** Deletar texto selecionado ao digitar
#+begin_src emacs-lisp
(delete-selection-mode t)
#+end_src

** Destaque da linha corrente
#+begin_src emacs-lisp
(global-hl-line-mode 1)
#+end_src

** Quebra de linha visual
#+begin_src emacs-lisp
(global-visual-line-mode t)
#+end_src

** Carregar alterações externas
#+begin_src emacs-lisp
(global-auto-revert-mode t)
#+end_src

** Prompt de confirmação
#+begin_src emacs-lisp
(fset 'yes-or-no-p 'y-or-n-p)
#+end_src

** Habilita o fechamento de pares
#+begin_src emacs-lisp
(electric-pair-mode 1)
#+end_src

** Sai do minibuffer com ESC
#+begin_src emacs-lisp
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
#+end_src

** Atalhos para zoom
#+begin_src emacs-lisp
(global-set-key (kbd "C-=") 'text-scale-increase) 
(global-set-key (kbd "C--") 'text-scale-decrease)
#+end_src

* Sistema
** Backup
#+begin_src emacs-lisp
(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))
#+end_src

** Custom settings
#+begin_src emacs-lisp
(setq custom-file (locate-user-emacs-file "custom.el"))
(load custom-file 'noerror 'nomessage)
#+end_src
* Fontes 
#+begin_src emacs-lisp
(set-face-attribute 'default nil :font "Anonymice Pro Nerd Font Bold Italic 13")
(set-face-attribute 'variable-pitch nil :font "Anonymice Pro Nerd Font Bold Italic 13")
(set-face-attribute 'fixed-pitch nil :font "Anonymice Pro Nerd Font Bold Italic 13")

#+end_src
* Gerenciamento ds pacotes
** Configuração do package
#+begin_src emacs-lisp
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
#+end_src

** Configuração do use-package
#+begin_src emacs-lisp
(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)
#+end_src

* Pacotes
** Evil Mode
#+begin_src emacs-lisp
(use-package evil
  :init
  (setq evil-want-integration t
        evil-want-keybinding nil
        evil-vsplit-window-right t
        evil-split-window-below t
        evil-undo-system 'undo-redo)
  (evil-mode))

(use-package evil-collection
  :after evil
  :config
  (add-to-list 'evil-collection-mode-list 'help) ;; evilify help mode
  (evil-collection-init))
#+end_src

** Vertico
#+begin_src emacs-lisp
(use-package vertico
  :bind (:map vertico-map
              ("C-j" . vertico-next)
              ("C-k" . vertico-previous)
              ("C-f" . vertico-exit)
              :map minibuffer-local-map
              ("M-h" . backward-kill-word))
  :custom
  (vertico-cycle t)
  :init
  (vertico-mode))
#+end_src

** Marginalia
;; Informações adicionais nas seleções.
#+begin_src emacs-lisp
(use-package marginalia
  :after vertico
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
  :init
  (marginalia-mode))
#+end_src
* Abrir Terminal
** Orderless
;; Busca fuzzy no minibuffer.
#+begin_src emacs-lisp
(use-package orderless
  :config
  (setq completion-styles '(orderless basic)))
#+end_src

** Consult
;; Comandos para busca e navegação.

    - =consult-grep=: Grep recursivo.
    - =consult-find=: Busca recursiva de arquivos.
    - =consult-outline=: Busca entre cabeçalhos (outline).
    - =consult-line=: Busca entre as linhas do buffer.
    - =consult-buffer=: Circula entre os buffers abertos.

#+begin_src emacs-lisp
(use-package consult)
#+end_src

** Which Key
#+begin_src emacs-lisp
(use-package which-key
  :init
  (which-key-mode 1)
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-allow-imprecise-window-fit nil
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 6
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.8
        which-key-max-description-length 25
        which-key-separator " → "))
#+end_src

** Doom Themes
#+begin_src emacs-lisp
(use-package doom-themes
  :config
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t)
  ;; Sets the default theme to load!!!
  (load-theme 'doom-moonlight t)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))
#+end_src
** Doom Modeline
#+begin_src emacs-lisp
(use-package doom-modeline
  :ensure t
  :hook
  (after-init . doom-modeline-mode)
  :custom
  (set-face-attribute 'mode-line nil :font "Anonymice Pro Nerd Font Bold Italic" :height 110) 
  (set-face-attribute 'mode-line-inactive nil :font "Anonymice Pro Nerd Font Bold Italic" :height 110) 
  :config
  (setq doom-modeline-enable-word-count t))

#+end_src
** Counsel
#+begin_src emacs-lisp
(use-package counsel
  :config
  (setq ivy-extra-directories nil))
#+end_src>
* Diminish
#+begin_src emacs-lisp
(use-package diminish)
#+end_src

* Company
;; Adicione configuração aqui.
#+begin_src emacs-lisp
(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t)) 
(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))
#+end_src
* Projecfile
#+begin_src emacs-lisp
(use-package projectile
  :diminish projectile-mode
  :config
  (projectile-mode))
#+end_src
* Funções
** Recarga das configurações 
#+begin_src emacs-lisp
(defun ze/reload-settings () 
  (interactive)
  (load-file "~/.config/emacs/init.el")) 
#+end_src
** Abrir config.org 
#+begin_src emacs-lisp
(defun ze/open-emacs-config () 
  (interactive)
  (find-file "~/.config/emacs/config.org")) 
#+end_src
** Abrir diretório de configurações pessoais 
#+begin_src emacs-lisp
  (defun ze/emacs-personal-files () 
    (interactive)
    (let((default-directory "~/.config/emacs"))
       (call-interactively 'find-file))) 
#+end_src

* General (atalhos)
#+begin_src emacs-lisp
(use-package general 
  :config
  (general-evil-setup)
  ;; set up 'SPC' as the global leader key
  (general-create-definer ze/leader-keys
    :states '(normal insert visual emacs) 
    :keymaps 'override
    :prefix "SPC" ;; set leader
    :global-prefix "M-SPC") ;; access leader in insert mode

  (ze/leader-keys
    "TAB TAB" '(comment-line :wk "Comment lines")) 

  ;; Buffer/bookmarks
  (ze/leader-keys
    "b" '(:ignore t :wk "Buffers/Bookmarks")
    "b b" '(switch-to-buffer :wk "Switch to buffer")
    "b i" '(ibuffer :wk "Ibuffer")
    "b k" '(kill-current-buffer :wk "Kill current buffer")
    "b s" '(basic-save-buffer :wk "Save buffer")
    "b l" '(list-bookmarks :wk "List bookmarks")
    "b m" '(bookmark-set :wk "Set bookmark")
    "q q" '(save-buffers-kill-terminal :wk "Quit emacs"))

  ;; Files
  (ze/leader-keys
    "f" '(:ignore t :wk "Files") 
    "." '(find-file :wk "Find file") 
    "f f" '(find-file :wk "Find file")
    "f p" '(ze/emacs-personal-files :wk "Open personal config files")
    "f c" '(ze/open-emacs-config :wk "Open emacs config.org"))

  ;; Helpers
  (ze/leader-keys
    "h" '(:ignore t :wk "Helpers")
    "h r r" '(ze/reload-settings :wk "Reload emacs settings"))) 
#+end_src

* Org Mode
**  Correção para fechamento de pares 
 Evita completar =<>= em org-mode para não atrapalhar os snippets de templates.
#+begin_src emacs-lisp
(add-hook 'org-mode-hook (lambda ()
           (setq-local electric-pair-inhibit-predicate
                   `(lambda (c)
                  (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))
#+end_src
**  Endentação de blocos 
#+begin_src emacs-lisp
(setq org-edit-src-content-indentation 0) ; Zera a margem dos blocos
(electric-indent-mode -1)                 ; Desliga a indentação automática
#+end_src
**  Org Bullets 
#+begin_src emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode) 
(use-package org-bullets
:custom
(org-bullets-bullet-list '("▶" "▷" "◆" "◇" "▪" "▪" "▪")))  
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src
** Templates de Blocos
#+begin_src emacs-lisp
    (require 'org-tempo)
#+end_src
* Rainbow Delimiters
#+begin_src emacs-lisp
(use-package rainbow-delimiters
  :hook ((prog-mode . rainbow-delimiters-mode)
         (emacs-lisp-mode . rainbow-delimiters-mode)
         (clojure-mode . rainbow-delimiters-mode)))
#+end_src

*  Treemacs 
#+begin_src emacs-lisp
(use-package treemacs
  :bind
  (:map global-map
        ("M-\\" . treemacs))
  :config
  (setq treemacs-no-png-images t
        treemacs-is-never-other-window nil))
#+end_src

* Vterm
** DONE Dependências para compilar:

   - =cmake=
   - =libtool-bin=
   - =libvterm-dev=

#+begin_src emacs-lisp
(use-package vterm
:config
(setq shell-file-name "/bin/bash"
      vterm-max-scrollback 5000))
  ;; Define a leader key for vterm
  (ze/leader-keys
    "t" '(:ignore t :which-key "terminal")
    "tt" '(open-vterm-split :which-key "open vterm"))
(defun open-vterm-split ()
  (interactive)
  (split-window-below 20) ;; Adjust the number to set the height of the vterm window
  (other-window 1)
  (vterm))
#+end_src


* Suporte a Python
;; Usando as teclas GCC
** Usando o Evil-Commentary
#+begin_src emacs-lisp
(use-package evil-commentary
  :ensure t
  :config
  (evil-commentary-mode))
#+end_src

** Configuração do modo Python
#+begin_src emacs-lisp
(use-package python-mode
  :ensure t
  :hook (python-mode . lsp-deferred)
  :custom
  (python-shell-interpreter "python3"))

(use-package blacken
  :ensure t
  :hook (python-mode . blacken-mode)
  :custom
  (blacken-line-length 88))

(use-package lsp-mode
  :hook ((python-mode . lsp-deferred)
         (lsp-mode . lsp-enable-which-key-integration))
  :commands lsp
  :custom
  (electric-indent-mode 1)
(indent-region)
  (lsp-headerline-breadcrumb-enable nil)
  (lsp-idle-delay 0.1)
  (lsp-log-io nil)
  (lsp-completion-provider :capf)
  (lsp-prefer-capf t)
  (lsp-enable-symbol-highlighting nil)
  (lsp-enable-on-type-formatting nil)
  (lsp-enable-indentation nil)
  (lsp-enable-snippet nil)
  (lsp-keymap-prefix "C-c l")
  (lsp-enable-file-watchers t) ;; Habilitar file watchers
  (lsp-file-watch-threshold 1000) ;; Limite de arquivos para watchers
  (setq lsp-file-watch-ignored-directory '("[/\\\\]\\.git$"
                                        "[/\\\\]\\.venv$"
                                        "[/\\\\]node_modules$"
                                        "[/\\\\]\\.cask$"
                                        "[/\\\\]\\.cache$"
                                        "[/\\\\]\\.python-version$"
                                        "[/\\\\]\\.mypy_cache$"
                                        "[/\\\\]\\.pytest_cache$"
                                        "[/\\\\]\\.pyre$"
                                        "[/\\\\]\\.ruff_cache$")))


(use-package lsp-ui
  :commands lsp-ui-mode
  :custom
  (lsp-ui-doc-enable t)
  (lsp-ui-doc-header t)
  (lsp-ui-doc-include-signature t)
  (lsp-ui-doc-position 'at-point)
  (lsp-ui-doc-max-width 150)
  (lsp-ui-doc-max-height 30)
  (lsp-ui-sideline-enable t)
  (lsp-ui-sideline-show-hover t)
  (lsp-ui-sideline-show-diagnostics t)
  (lsp-ui-sideline-show-code-actions t)
  (lsp-ui-sideline-ignore-duplicate t))

(use-package lsp-ivy
  :commands lsp-ivy-workspace-symbol)
(use-package lsp-pyright
  :ensure t
  :hook (python-mode . (lambda ()
                         (require 'lsp-pyright)
                         (lsp-deferred)))
  :custom
  (lsp-pyright-python-executable-cmd "python3"))
#+end_src



emacs/history
;; -*- mode: emacs-lisp; coding: utf-8-unix -*-
;; Minibuffer history file, automatically generated by ‘savehist’.

(setq savehist-minibuffer-history-variables '(xref--read-pattern-history woman-topic-history org-link--history xref--read-identifier-history minibuffer-history buffer-name-history shell-command-history command-history consult--find-history evil-ex-history file-name-history extended-command-history))
(setq minibuffer-history '("yes" "pyright" "yes"))
(setq buffer-name-history '("config.org"))
(setq shell-command-history '("/bin/bash"))
(setq command-history '((evil-delete 442 484 'line) (evil-delete 485 505 'line) (evil-delete 484 505 'line) (evil-change 94 99 'inclusive) (evil-delete 1 28 'line) (evil-delete 1 2 'line) (evil-delete 1 2 'line) (evil-delete 1 2 'line) (evil-delete 1 2 'line) (evil-delete 1 18 'line) (evil-delete 426 449 'line) (evil-delete 32 61 'line) (evil-replace 168 169 'exclusive 32) (evil-delete 139 168 'line) (evil-delete 168 195 'line) (evil-goto-mark-line 39) (evil-replace 470 471 'exclusive 105) (evil-set-marker 97) (evil-delete 176 192 'inclusive) (evil-delete 60 87 'line) (find-file "~/Cuso_Python/Gerador_de_senha/dataclasse2.py" t) (evil-delete 233 248 'line) (find-file "~/Cuso_Python/Gerador_de_senha/> tipos.py" t) (evil-delete 100 108 'line) (evil-delete 100 120 'line) (evil-delete 100 121 'line) (evil-delete 112 114 'line) (evil-delete 112 123 'line) (evil-delete 112 127 'line) (evil-delete 115 115 'line) (evil-delete 116 116 'line) (ze/emacs-personal-files) (execute-extended-command nil "ze/emacs-personal-files" "/") (evil-replace 164 165 'exclusive 60) (evil-replace 164 165 'exclusive 33) (evil-delete 116 132 'line) (evil-delete 208 209 'line) (evil-delete 208 209 'line) (evil-replace 187 188 'exclusive 114) (evil-delete 171 180 'line) (evil-delete 88 103 'line) (evil-delete 103 120 'line) (evil-replace 69 70 'exclusive 120) (find-file "~/Cuso_Python/algoritimo/soma.py" t) (evil-find-char-to nil 116) (evil-find-char-to nil 110) (evil-delete 1 47 'line) (evil-delete 48 60 'line) (evil-delete 61 78 'line) (evil-delete 19 79 'line) (evil-delete 80 97 'line) (evil-replace 96 97 'exclusive 100) (evil-delete 7123 7124 'line) (evil-delete 7123 7124 'line) (evil-delete 7208 7209 'line) (evil-delete 7208 7209 'line) (evil-delete 7208 7209 'line) (evil-delete 9783 9784 'line) (evil-delete 9783 9784 'line) (evil-delete 9806 9808 'line) (evil-delete 9785 9786 'line) (evil-find-char-to nil 107) (evil-delete 10628 10696 'line) (evil-delete 10601 10651 'line) (evil-delete 5080 5084 'line) (evil-replace 825 826 'exclusive 83) (evil-replace 825 826 'exclusive 82) (evil-replace 825 826 'exclusive 114) (evil-record-macro 113) (evil-delete 774 782 'line) (evil-find-char nil 100) (evil-find-char nil 100) (evil-find-char nil 100) (evil-find-char-to nil 104) (evil-record-macro 113) (evil-find-char nil 97) (evil-replace 4456 4457 'exclusive 109) (evil-replace 2349 2350 'exclusive 109) (evil-replace 2305 2306 'exclusive 109) (evil-replace 2242 2243 'exclusive 109) (evil-replace 2178 2179 'exclusive 109) (evil-replace 2212 2213 'exclusive 109) (evil-delete 2454 2455 'line) (evil-delete 359 388 'line) (evil-delete 132 133 'line) (evil-delete 4295 4300 'line) (evil-replace 4159 4160 'exclusive 49) (evil-replace 4133 4134 'exclusive 49) (evil-replace 4108 4109 'exclusive 56) (evil-replace 4082 4083 'exclusive 49) (evil-replace 4056 4057 'exclusive 49) (evil-replace 4031 4032 'exclusive 56) (evil-find-char nil 53) (evil-replace 4139 4140 'exclusive 56) (evil-replace 4115 4116 'exclusive 55) (evil-replace 4091 4092 'exclusive 54) (evil-replace 4067 4068 'exclusive 53) (evil-replace 4043 4044 'exclusive 52) (evil-replace 4019 4020 'exclusive 51) (evil-replace 3995 3996 'exclusive 50)))
(setq consult--find-history '("#"))
(setq evil-ex-history '("q" ".!/etc/shells" ".!/etc/shells/" ".!/etc/shell" ".!cat" "q" "w" "q" "w" "wq"))
(setq file-name-history '("~/Cuso_Python/Gerador_de_senha" "~/Cuso_Python/Gerador_de_senha/dataclasse2.py" "~/Cuso_Python/Gerador_de_senha/> tipos.py" "~/Cuso_Python/Gerador_de_senha" "~/Cuso_Python/algoritimo/soma.py" "~/Cuso_Python/algoritimo" "~/Cuso_Python/Gerador_de_senha/gorjeta.py" "~/Cuso_Python/Gerador_de_senha" "~/.config/emacs/config.org" "~/.config/" "~/.config/emacs/config.org" "~/Cuso_Python/Gerador_de_senha" "~/Cuso_Python/Gerador_de_senha/main.py" "~/Automacao" "~/classe.py" "~/.config/emacs/config-bkp.org" "~/classe.py" "~/test.py" "~/polygon.py" "~/aula.py" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/dist.py" "~/exmplo.py" "~/recursivo.py" "~/polygon.py" "~/mypoligon.py" "~/.config/" "~/.config/emacs/" "~/.config/emacs/config.org" "~/github/fast_zero/.github/workflows/pipeline.yaml" "~/github/fast_zero/.github/workflows/*scratch*" "~/github/fast_zero/.github" "~/.config/emacs/config.org" "~/github/fast_zero/tests/conftest.py" "~/github/fast_zero/tests/test_app.py" "~/github/fast_zero/tests/test_db.py" "~/github/fast_zero/fast_zero/app.py" "~/exemplo.org" "~/dados.txt" "~/exemplo.org" "~/exemplo.txt" "~/exmplo.py" "~/exemplo.txt" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org" "~/.config/emacs/init.el" "~/.config/emacs/config.org"))

emacs/config.el
(org-indent-mode t)

;;(setq org-edit-src-content-indentation 0) ; Zera a margem dos blocos
;;(electric-indent-mode -1)                 ; Desliga a indentação automática

(setq inhibit-startup-message t)

(menu-bar-mode -1)            ; no menu bar
(tool-bar-mode -1)            ; no tools bar
(scroll-bar-mode -1)          ; no scroll bars
(tooltip-mode -1)             ; no tooltips
(set-fringe-mode 10)          ; frame edges set to 10px
(column-number-mode 1)        ; modeline shows column number
(save-place-mode 1)           ; remember cursor position
(recentf-mode 1)              ; remember recent files
(savehist-mode 1)             ; enable history saving

(global-display-line-numbers-mode 1) 
(setq display-line-numbers-type 'relative) 
(dolist (mode '(org-mode-hook
                vterm-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

(setq-default indent-tabs-mode nil)

(delete-selection-mode t)

(global-hl-line-mode 1)

(global-visual-line-mode t)

(global-auto-revert-mode t)

(fset 'yes-or-no-p 'y-or-n-p)

(electric-pair-mode 1)

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

(global-set-key (kbd "C-=") 'text-scale-increase) 
(global-set-key (kbd "C--") 'text-scale-decrease)

(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))

(setq custom-file (locate-user-emacs-file "custom.el"))
(load custom-file 'noerror 'nomessage)

(set-face-attribute 'default nil :font "Anonymice Pro Nerd Font Bold Italic 13")
(set-face-attribute 'variable-pitch nil :font "Anonymice Pro Nerd Font Bold Italic 13")
(set-face-attribute 'fixed-pitch nil :font "Anonymice Pro Nerd Font Bold Italic 13")

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))

(unless (package-installed-p 'use-package)
  (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(use-package evil
  :init
  (setq evil-want-integration t
        evil-want-keybinding nil
        evil-vsplit-window-right t
        evil-split-window-below t
        evil-undo-system 'undo-redo)
  (evil-mode))

(use-package evil-collection
  :after evil
  :config
  (add-to-list 'evil-collection-mode-list 'help) ;; evilify help mode
  (evil-collection-init))

(use-package vertico
  :bind (:map vertico-map
              ("C-j" . vertico-next)
              ("C-k" . vertico-previous)
              ("C-f" . vertico-exit)
              :map minibuffer-local-map
              ("M-h" . backward-kill-word))
  :custom
  (vertico-cycle t)
  :init
  (vertico-mode))

(use-package marginalia
  :after vertico
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
  :init
  (marginalia-mode))

(use-package orderless
  :config
  (setq completion-styles '(orderless basic)))

(use-package consult)

(use-package which-key
  :init
  (which-key-mode 1)
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-allow-imprecise-window-fit nil
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 6
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.8
        which-key-max-description-length 25
        which-key-separator " → "))

(use-package doom-themes
  :config
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t)
  ;; Sets the default theme to load!!!
  (load-theme 'doom-moonlight t)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))

(use-package doom-modeline
  :ensure t
  :hook
  (after-init . doom-modeline-mode)
  :custom
  (set-face-attribute 'mode-line nil :font "Anonymice Pro Nerd Font Bold Italic" :height 110) 
  (set-face-attribute 'mode-line-inactive nil :font "Anonymice Pro Nerd Font Bold Italic" :height 110) 
  :config
  (setq doom-modeline-enable-word-count t))

(use-package diminish)

(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t)) 
(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))

(use-package projectile
  :diminish projectile-mode
  :config
  (projectile-mode))

(defun ze/reload-settings () 
  (interactive)
  (load-file "~/.config/emacs/init.el"))

(defun ze/open-emacs-config () 
  (interactive)
  (find-file "~/.config/emacs/config.org"))

(defun ze/emacs-personal-files () 
  (interactive)
  (let((default-directory "~/.config/emacs"))
     (call-interactively 'find-file)))

(use-package general 
  :config
  (general-evil-setup)
  ;; set up 'SPC' as the global leader key
  (general-create-definer ze/leader-keys
    :states '(normal insert visual emacs) 
    :keymaps 'override
    :prefix "SPC" ;; set leader
    :global-prefix "M-SPC") ;; access leader in insert mode

  (ze/leader-keys
    "TAB TAB" '(comment-line :wk "Comment lines")) 

  ;; Buffer/bookmarks
  (ze/leader-keys
    "b" '(:ignore t :wk "Buffers/Bookmarks")
    "b b" '(switch-to-buffer :wk "Switch to buffer")
    "b i" '(ibuffer :wk "Ibuffer")
    "b k" '(kill-current-buffer :wk "Kill current buffer")
    "b s" '(basic-save-buffer :wk "Save buffer")
    "b l" '(list-bookmarks :wk "List bookmarks")
    "b m" '(bookmark-set :wk "Set bookmark")
    "q q" '(save-buffers-kill-terminal :wk "Quit emacs"))

  ;; Files
  (ze/leader-keys
    "f" '(:ignore t :wk "Files") 
    "." '(find-file :wk "Find file") 
    "f f" '(find-file :wk "Find file")
    "f p" '(ze/emacs-personal-files :wk "Open personal config files")
    "f c" '(ze/open-emacs-config :wk "Open emacs config.org"))

  ;; Helpers
  (ze/leader-keys
    "h" '(:ignore t :wk "Helpers")
    "h r r" '(ze/reload-settings :wk "Reload emacs settings")))

(add-hook 'org-mode-hook (lambda ()
           (setq-local electric-pair-inhibit-predicate
                   `(lambda (c)
                  (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))

(setq org-edit-src-content-indentation 0) ; Zera a margem dos blocos
(electric-indent-mode -1)                 ; Desliga a indentação automática

(add-hook 'org-mode-hook 'org-indent-mode) 
(use-package org-bullets
:custom
(org-bullets-bullet-list '("▶" "▷" "◆" "◇" "▪" "▪" "▪")))  
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(require 'org-tempo)

(use-package rainbow-delimiters
  :hook ((prog-mode . rainbow-delimiters-mode)
         (emacs-lisp-mode . rainbow-delimiters-mode)
         (clojure-mode . rainbow-delimiters-mode)))

(use-package treemacs
  :bind
  (:map global-map
        ("M-\\" . treemacs))
  :config
  (setq treemacs-no-png-images t
        treemacs-is-never-other-window nil))

(use-package vterm
:config
(setq shell-file-name "/bin/bash"
      vterm-max-scrollback 5000))
  ;; Define a leader key for vterm
  (ze/leader-keys
    "t" '(:ignore t :which-key "terminal")
    "tt" '(open-vterm-split :which-key "open vterm"))
(defun open-vterm-split ()
  (interactive)
  (split-window-below 20) ;; Adjust the number to set the height of the vterm window
  (other-window 1)
  (vterm))

(use-package evil-commentary
  :ensure t
  :config
  (evil-commentary-mode))

(use-package python-mode
  :ensure t
  :hook (python-mode . lsp-deferred)
  :custom
  (python-shell-interpreter "python3"))

(use-package blacken
  :ensure t
  :hook (python-mode . blacken-mode)
  :custom
  (blacken-line-length 88))

(use-package lsp-mode
  :hook ((python-mode . lsp-deferred)
         (lsp-mode . lsp-enable-which-key-integration))
  :commands lsp
  :custom
  (electric-indent-mode 1)
(indent-region)
  (lsp-headerline-breadcrumb-enable nil)
  (lsp-idle-delay 0.1)
  (lsp-log-io nil)
  (lsp-completion-provider :capf)
  (lsp-prefer-capf t)
  (lsp-enable-symbol-highlighting nil)
  (lsp-enable-on-type-formatting nil)
  (lsp-enable-indentation nil)
  (lsp-enable-snippet nil)
  (lsp-keymap-prefix "C-c l")
  (lsp-enable-file-watchers t) ;; Habilitar file watchers
  (lsp-file-watch-threshold 1000) ;; Limite de arquivos para watchers
  (setq lsp-file-watch-ignored-directory '("[/\\\\]\\.git$"
                                        "[/\\\\]\\.venv$"
                                        "[/\\\\]node_modules$"
                                        "[/\\\\]\\.cask$"
                                        "[/\\\\]\\.cache$"
                                        "[/\\\\]\\.python-version$"
                                        "[/\\\\]\\.mypy_cache$"
                                        "[/\\\\]\\.pytest_cache$"
                                        "[/\\\\]\\.pyre$"
                                        "[/\\\\]\\.ruff_cache$")))


(use-package lsp-ui
  :commands lsp-ui-mode
  :custom
  (lsp-ui-doc-enable t)
  (lsp-ui-doc-header t)
  (lsp-ui-doc-include-signature t)
  (lsp-ui-doc-position 'at-point)
  (lsp-ui-doc-max-width 150)
  (lsp-ui-doc-max-height 30)
  (lsp-ui-sideline-enable t)
  (lsp-ui-sideline-show-hover t)
  (lsp-ui-sideline-show-diagnostics t)
  (lsp-ui-sideline-show-code-actions t)
  (lsp-ui-sideline-ignore-duplicate t))

(use-package lsp-ivy
  :commands lsp-ivy-workspace-symbol)
(use-package lsp-pyright
  :ensure t
  :hook (python-mode . (lambda ()
                         (require 'lsp-pyright)
                         (lsp-deferred)))
  :custom
  (lsp-pyright-python-executable-cmd "python3"))
